include/range/v3/view/for_each.hpp

```diff
@@ -58,35 +46,11 @@ namespace ranges
                 )
             public:
                 template<typename Rng, typename F>
-                using Concept = meta::and_<
-                    Range<Rng>,
-                    IndirectInvocable<F, iterator_t<Rng>>,
-                    Range<concepts::Invocable::result_t<F&, range_common_reference_t<Rng>>>>;
-
-                template<typename Rng, typename F,
-                    CONCEPT_REQUIRES_(Concept<Rng, F>())>
-                for_each_view<all_t<Rng>, F> operator()(Rng && rng, F f) const
-                {
-                    return {all(std::forward<Rng>(rng)), std::move(f)};
-                }
-
-            #ifndef RANGES_DOXYGEN_INVOKED
-                // For better error reporting
-                template<typename Rng, typename F,
-                    CONCEPT_REQUIRES_(!Concept<Rng, F>())>
-                void operator()(Rng &&, F) const
-                {
-                    CONCEPT_ASSERT_MSG(Range<Rng>(),
-                        "Rng is not a model of the Range concept.");
-                    CONCEPT_ASSERT_MSG(IndirectInvocable<F, iterator_t<Rng>>(),
-                        "The function F is not callable with arguments of the type of the range's "
-                        "common reference type.");
-                    CONCEPT_ASSERT_MSG(Range<concepts::Invocable::result_t<F&,
-                        range_common_reference_t<Rng>>>(),
-                        "To use view::for_each, the function F must return a model of the Range "
-                        "concept.");
-                }
-            #endif
+                auto operator()(Rng && rng, F f) const
+                RANGES_DECLTYPE_AUTO_RETURN
+                (
+                    join(transform(std::forward<Rng>(rng), std::move(f)))
+                )
```

Comment:
I'd like to do this without losing the error messages. I think the best way to do that is something like:
```c++
struct for_each_fn
{
private:
    template <class Rng, class F>
    static auto impl(int, Rng &&rng, F f)
    RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
    (
        join(transform(std::forward<Rng>(rng), std::move(f)))
    )
    template <class Rng, class F>
    static void impl(long, Rng &&, F)
    {
        // Like the pre-existing operator() with the CONCEPT_ASSERT_MSGs,
        // but updated to the actual current requirements of join and transform.
        // These messages may get out of date, but the view will keep working.
    }
public:
    auto operator()(Rng &&rng, F f) const
    RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
    (
        for_each_fn::impl(42, std::forward<Rng>(rng), std::move(f))
    )
};
```